File: 2024/Day5/1.py
def parse_rules(rule_lines):
    rules = {}
    for rule in rule_lines:
        x, y = map(int, rule.split("|"))
        rules[(x, y)] = True  # X must come before Y
    return rules

def is_valid_update(update, rules):
    for i in range(len(update)):
        for j in range(i + 1, len(update)):
            x, y = update[i], update[j]
            if (y, x) in rules:  # If there's a rule Y|X, this order is invalid
                return False
    return True

def find_middle_page(update):
    return update[len(update) // 2]

def main():
    with open("input.txt", "r") as file:
        lines = file.readlines()

    # Separate rules and updates
    rules_section = []
    updates_section = []
    parsing_rules = True
    for line in lines:
        line = line.strip()
        if not line:
            parsing_rules = False
            continue
        if parsing_rules:
            rules_section.append(line)
        else:
            updates_section.append(line)

    # Parse rules
    rules = parse_rules(rules_section)

    # Process updates
    total_middle_sum = 0
    for update_line in updates_section:
        update = list(map(int, update_line.split(",")))
        if is_valid_update(update, rules):
            total_middle_sum += find_middle_page(update)

    print(f"Total Middle Sum: {total_middle_sum}")

if __name__ == "__main__":
    main()

File: 2024/Day5/1.swift
import Foundation

let filePath = "input.txt"

struct Rule: Hashable {
    let x: Int
    let y: Int
}

func parseRules(_ ruleLines: [String]) -> Set<Rule> {
    var rules = Set<Rule>()
    for rule in ruleLines {
        let parts = rule.split(separator: "|").compactMap { Int($0) }
        if parts.count == 2 {
            rules.insert(Rule(x: parts[0], y: parts[1]))
        }
    }
    return rules
}

func isValidUpdate(_ update: [Int], rules: Set<Rule>) -> Bool {
    for i in 0..<update.count {
        for j in (i + 1)..<update.count {
            let x = update[i], y = update[j]
            if rules.contains(Rule(x: y, y: x)) { // If there's a rule Y|X, this order is invalid
                return false
            }
        }
    }
    return true
}

func findMiddlePage(_ update: [Int]) -> Int {
    return update[update.count / 2]
}

do {
  let content = try String(contentsOfFile: filePath, encoding: .utf8)

  let lines = content.split(separator: "\n").map { String($0) }

  var rulesSection = [String]()
  var updatesSection = [String]()
  var parsingRules = true

  for line in lines {
    if line.contains(",") {
      parsingRules = false
    }
    if parsingRules {
      rulesSection.append(line)
    } else {
      updatesSection.append(line)
    }
  }

  // Parse rules
  let rules = parseRules(rulesSection)

  var totalMiddleSum = 0
  for updateLine in updatesSection {
        let update = updateLine.split(separator: ",").compactMap { Int($0) }
        if isValidUpdate(update, rules: rules) {
            totalMiddleSum += findMiddlePage(update)
        }
  }

  print("Total Middle Sum: \(totalMiddleSum)")

} catch {
  print("Error: \(error)")
}
